// www/index.js
// www module, served on /

// happner defaults for this module

module.exports.$happner = {
  config: {
    component: {
      startMethod: 'start',
      stopMethod: 'stop',
      schema: {
        exclusive: true,
        methods: {
          'start': {},
          'stop': {},
          'exampleExchangeMethod': {
            alias: 'example' // called from client in client.exchange.www.example()
          }
        }
      },
      web: {
        routes: {
          static: 'static'
        }
      },
      events: {
        ping: {},
        'wildcard/*': {}
      },
      data: {
        'storage/*': {}
      }
    }
  }
}

module.exports.exampleExchangeMethod = function($happn, arg, callback) {

  callback(null, arg.true);

}

module.exports.start = function($happn, callback) {
  // $happn.log.info('Starting www component.');

  var terminal, address, url;

  address = $happn.info.datalayer.address;
  url = 'http://'+address.address+':'+address.port;

  // If present, add 'www' command to terminal.

  if (terminal = $happn.exchange.terminal) {                   // TODO: for various browsers: open in same/existing tab
    terminal.register('www', {                                // 
      description: '(osx) Open browser to the www module.',  //
      run: function(args, callback) {                       //
        require('child_process').exec('open ' + url, function(e) {
          callback(e);
        });
      }
    });
  }

  var seq1 = 0;
  setInterval(function() {
    seq1++
    $happn.emit('ping', {seq: seq1});
  }, 100);

  var seq2 = 0;
  setInterval(function() {
    seq2++;
    $happn.emit('wildcard/' + seq2, {seq: seq2});
  }, 100);

  callback(null);
}

module.exports.stop = function($happn, callback) {
  // $happn.log.info('Stopping www component.');
  callback(null);
}
